cisc/certificate.go

```diff
@@ -0,0 +1,321 @@
+		log.Info("No change in the certificate")
+
+	} else {
+		log.Info("The new certificate is different :")
+		if !((cert1.NotAfter).Equal(cert2.NotAfter)) {
+			log.Info("    Expiration date modification: " + cert1.NotAfter.String() + " to " + cert2.NotAfter.String())
+		}
+		if !((cert1.Subject.CommonName) == (cert2.Subject.CommonName)) {
+			log.Info("    Subject modification: " + cert1.Subject.CommonName + " to " + cert2.Subject.CommonName)
+		}
+		if !((cert1.Issuer.CommonName) == (cert2.Issuer.CommonName)) {
+			log.Info("    Issuer modification: " + cert1.Issuer.CommonName + " to " + cert2.Issuer.CommonName)
+		}
+
+		if !((cert1.PublicKey.(*rsa.PublicKey).N).Cmp(cert2.PublicKey.(*rsa.PublicKey).N) == 0) {
+			log.Info("    PublicKey modification")
+		}
+	}
+}
+
+func pemToCertificate(certPem string) (*x509.Certificate, error) {
+	block, _ := pem.Decode([]byte(certPem))
+	if block == nil {
+		return nil, errors.New("Fail decode pem")
+	}
+	cert, err := x509.ParseCertificate(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+func isCert(cert string) bool {
+	_, err := pemToCertificate(cert)
+	return err == nil
+}
+func check(cert string) bool {
+	rootPool := x509.NewCertPool()
+	certRoots, err := pemToCertificate(LetsencryptCert)
+	if err != nil {
+		log.Info("The root cert is not valid")
+	}
+	rootPool.AddCert(certRoots)
+	certCheck, err := pemToCertificate(cert)
+	if err != nil {
+		log.Info("The checked cert is not valid")
+	}
+	opts := x509.VerifyOptions{
+		DNSName: certCheck.DNSNames[0],
+		Roots:   rootPool,
+	}
+	if _, err := certCheck.Verify(opts); err != nil {
+		log.Info("certificate not valid: " + err.Error())
+		return false
+	}
+	log.Info("Cert OK")
+	return true
+
+}
+func renewCert(cert string) string {
+
+	oldcert, err := pemToCertificate(cert)
+	if err != nil {
+		log.Fatal(err)
+	}
+	//get the serial number and compute the URI
+	serial := fmt.Sprintf("%x", oldcert.SerialNumber)
+	certuri := "https://acme-v01.api.letsencrypt.org/acme/cert/0" + serial
+	// Create a client to the acme of letsencrypt.
+	cli, err := letsencrypt.NewClient("https://acme-v01.api.letsencrypt.org/directory")
+	if err != nil {
+		log.Fatal(err)
+	}
+	//renew the certificate
+	certRen, err := cli.RenewCertificate(string(certuri))
+	if err != nil {
+		log.Fatal("Problem while renewing Certificate:" + err.Error())
+	}
+	//write in the memory
+	fullchain, err := cli.Bundle(certRen)
+	if err != nil {
+		log.Fatal("can't bundles the certificate", err)
+	}
+	ioutil.WriteFile("fullchain.pem", fullchain, 0644)
+	log.Info("Certificate successfully renewed")
+
+	return string(fullchain)
+}
+
+func revokeCert(path string, cert string) {
+	// Create a client to the acme of letsencrypt.
+	log.Info("Revoking the key")
+	cli, err := letsencrypt.NewClient("https://acme-v01.api.letsencrypt.org/directory")
+	//cli, err := letsencrypt.NewClient("https://acme-staging.api.letsencrypt.org/directory")
+	if err != nil {
+		log.Fatal(err)
```

Comment:
Don't use `log.Fatal` in methods. Return an error and let the caller decide if he wants to return the error higher up.